LOGIN_FLOW.md

```diff
@@ -0,0 +1,355 @@
+app to be open in different tabs, where each tab could be open in a different
+group or chatroom.
+
+#### Keeping the shared part of the local state in sync
+
+When the shared state changes, the service worker broadcasts an event, as
+discussed. Besides this, the service worker broadcasts a range of events to tabs
+other than simply 'the state changed'. For example, the `LOGIN` and `LOGOUT`
+events are other events that are conveyed to let individual tabs know that the
+current user session changed. In general terms, it can be said that Group Income
+(the app) is event-driven as a consequence of using a service worker architecture,
+and that presentation depends on correctly sending and processing events.
+
+## Login process
+
+Since logging in and out occurs in the browsing context as a result of user
+interaction, it's special in the sense that it's a 'shared state' event that
+originates in the browsing context. This makes the process be harder to picture
+than most other things, where there is a clearer hierarchy.
+
+In addition, there are two entry points to the log in process: one that happens
+as a result of an explicit user action (e.g., logging in) and another one that
+happens when the app is loaded in a new tab (i.e., loading Group Income in a new
+tab). The latter is so that new tabs can re-use an existing session and users
+are not prompted to log in again when they have already done so in the past.
+
+### Login events
+
+#### Overview
+
+  * `LOGIN`: The `LOGIN` event represents an active login intent, which is to
+   say, it represents that there is an active login process, which may either
+   fail or succeed. This event can originate in the service worker or in a local
+   browsing context.
+  * `LOGIN_COMPLETE`: The `LOGIN_COMPLETE` is emitted following a `LOGIN` event
+  and signifies that the login process completed successfully. This event is
+  only local, which means that it doesn't originate on the service worker. Its
+  purpose is to let the app know that an active user session exists and has been
+  initialised.
+  * `LOGIN_ERROR`: The `LOGIN_ERROR` is similar to the `LOGIN_COMPLETE` error,
+  except that it means that the login process did _not_ complete successfully,
+  and, as a result, there is no active user session.
+  * `LOGOUT`: The `LOGOUT` event represents the end of an active user session.
+  This event originates in the service worker.
+
+#### `LOGIN` event
+
+##### Triggers
+
+###### `actions/identity.js`
+
+The `LOGIN` event is emitted in the service worker after calling
+`'gi.actions/identity/login'`. Its purpose is to let all open tabs that a new
+active session exists.
+
+Note that there is no corresponding `LOGIN_ERROR` emitted here, because an error
+would be thrown directly for the caller to handler. The reason for this is that,
+while all tabs need to know that a user recently logged in, only the caller
+needs to know when it logging in fails (for example, to display feedback in the
+login form)
```

Comment:
The way this is written makes it sound like if the `LOGIN` event is emitted in the service worker, it is *not* emitted in the browsing context. Is that right? If so, it doesn't make much sense because down below for the `app` section it says:

> The purpose of
emitting this event is to trigger the various handlers that set up the app state
based on an existing session.

Wouldn't these handlers still need to get triggered when logging in the for the first time?

Can this section be updated to address this?

Reply:
That's right. `LOGIN` can happen in the SW (when using a password) _(x)or_ in the browsing context (when an active session exists).